### ISSUE TITLE
Changes in CVMFS breaking verification

### ISSUE THREAD

**Issue by colsond** [hep-gc/shoal#101]
I noticed over the weekend that the shoal server stopped verifying squids. I thought something went wrong in the code and the verification thread crashed but that was not the case. They seem to have discontinued the following URL:

```
http://cvmfs02.grid.sinica.edu.tw:8000
```

Since the verification thread blacklists any squid that fails even one URL this was causing all of the squids to fail verification. I will be making a pull request soon with a patch that fixes this but I think we need some way to detect these sort of errors because it wasn't trivial to track down what was going wrong. @rptaylor do you know why this URL has been discontinued and can you think of a reason [any of the other following URLs](https://github.com/hep-gc/shoal/blob/verification/shoal-server/shoal_server/config.py#L48) will be discontinued soon?

```
paths = [
"http://cvmfs.racf.bnl.gov:8000/cvmfs/atlas/.cvmfswhitelist",
"http://cvmfs.racf.bnl.gov:8000/cvmfs/atlas-condb/.cvmfswhitelist",
"http://cvmfs.racf.bnl.gov:8000/cvmfs/sft/.cvmfswhitelist",
"http://cvmfs.racf.bnl.gov:8000/cvmfs/grid/.cvmfswhitelist",
"http://cvmfs.fnal.gov:8000/cvmfs/atlas/.cvmfswhitelist",
"http://cvmfs.fnal.gov:8000/cvmfs/atlas-condb/.cvmfswhitelist",
"http://cvmfs.fnal.gov:8000/cvmfs/sft/.cvmfswhitelist",
"http://cvmfs.fnal.gov:8000/cvmfs/grid/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/atlas.cern.ch/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/atlas-condb.cern.ch/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/sft.cern.ch/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/grid.cern.ch/.cvmfswhitelist",
"http://cernvmfs.gridpp.rl.ac.uk:8000/cvmfs/atlas/.cvmfswhitelist",
"http://cernvmfs.gridpp.rl.ac.uk:8000/cvmfs/atlas-condb/.cvmfswhitelist",
"http://cernvmfs.gridpp.rl.ac.uk:8000/cvmfs/sft/.cvmfswhitelist",
"http://cernvmfs.gridpp.rl.ac.uk:8000/cvmfs/grid/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/atlas/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/atlas-condb/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/sft/.cvmfswhitelist",
"http://cvmfs02.grid.sinica.edu.tw:8000/cvmfs/grid/.cvmfswhitelist"
]
```

After discussing this with Ron I think it would be a good idea to have a test proxy that we know will work. This would be used to test the URLs against and prune the list when one of them fails so we don't get across the board verification failure like we had this weekend.

**Comment by rptaylor** [comment]
The Taiwanese CVMFS server is still in production. Maybe for some reason they moved back to only using the /opt/ location instead of /cvmfs:
http://cvmfs02.grid.sinica.edu.tw:8000/opt
(Actually, I thought that was already the case... Are you sure the verification code didn't change?)

We can remove it from the verification list for now, but we should leave this issue open until we can figure out how to reintroduce the TW CVMFS server for verification later.

Changes to the CVMFS servers will probably almost never happen, so I don't think we need to put too much effort into a solution for that at this time.

**Comment by rptaylor** [comment]
This URL is available:
http://cvmfs02.grid.sinica.edu.tw:8000/opt/atlas.cern.ch/.cvmfswhitelist

Again, I'm pretty sure we already knew that the TW server wasn't updated to use the cvmfs/ directory instead of opt/ , but we just decided to leave it and wait until it was. So I'm not entirely convinced that something changed with the TW server.

**Comment by colsond** [comment]
Sorry I wasn't totally clear - it's not the entire TW server, it is just the bottom 4 (the /cvfms ones). These were 100% working until at least last Thursday. I've made a pull request for a patch that will remove them for the list.

**Comment by colsond** [comment]
I'm going to mark this resolved with the merge of #102 as Ryan thinks the urls will remain static enough to not merit dynamic testing and pruning of the list.

**Comment by rd37** [comment]
I thinks this has to be reopened ... if a static site goes down, then all squids fail.

**Comment by rptaylor** [comment]
We could add some smarter logic such as (just an example):
If all proxies fail on URLs for just one server, but can access other servers, the proxies are probably fine and the problem is probably with that server. CVMFS is already robust to this, so Shoal should be too, i.e. don't declare the proxy as having failed validation.
We could think about handling repo-specific failures in the same way.

But if just one proxy/server combination fails, it should probably be declared unverified.

If all proxy/server combinations fail, a human needs to look into it no matter what.

**Comment by rptaylor** [comment]
Actually I don't think the Taiwanese server ever stopped working -  #113 is probably the explanation (they started using a different port or started fully qualifying repo names or other change in format).

There are two issues: 
1. How do we update the verification URLs. That can be fixed now, see #113.
2. Smarter logic for handling verification failures. We could improve Shoal in this regard. I made a new issue #114 for this.

So now I'll close this.
